qualtran/bloqs/chemistry/black_boxes.py

```diff
@@ -0,0 +1,123 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Common Chemistry bloqs which have costs that differ from those currently implemented in qualtran.
+
+These are for temporary convenience to lock-in the quoted literature costs.
+"""
+from functools import cached_property
+from typing import Set, Tuple, TYPE_CHECKING
+
+import numpy as np
+from attrs import frozen
+from sympy import factorint
+
+from qualtran import Bloq, Register, Signature
+from qualtran.bloqs.basic_gates import Toffoli
+
+if TYPE_CHECKING:
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class PrepareUniformSuperposition(Bloq):
+    r"""Prepare a uniform superposition over $d$ basis states.
+
+    Uses quoted literature costs which relies on phase gradient for rotations.
+
+    Args:
+        d: The number of coefficients to prepare.
+        num_bits_rot_aa: The number of bits of precision for the single-qubit
+            rotation for amplitude amplification during the uniform state
+            preparataion. Default 8.
+
+    Registers:
+        d: the register to prepare the uniform superposition on.
+
+    Refererences:
+        [Even More Efficient Quantum Computations of Chemistry Through Tensor
+            hypercontraction](https://arxiv.org/abs/2011.03494) Page 39.
+    """
+    d: int
+    num_bits_rot_aa: int = 8
+
+    @cached_property
+    def signature(self) -> Signature:
+        regs = [Register('d', (self.num_non_zero - 1).bit_length())]
+        return Signature(regs)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        factors = factorint(self.d)
```

Comment:
can this get expensive?

Reply:
hmm, I haven't noticed it yet. This bloq should be pretty short lived so I'm not too worried.